### ISSUE TITLE
`Version` can be exhaustively matched

### ISSUE THREAD

**Issue by whataloadofwhat** [hyperium/http#155]
I feel like having `Version` be a load of `const`s rather than an `enum` was probably so that it could be extended without breakage. Unfortunately it seems `rustc` is too smart and sees through the ruse?

```rust
fn foo(version: Version) {
    match version {
        Version::HTTP_09 => http09(),
        Version::HTTP_10 => http10(),
        Version::HTTP_11 => http11(),
        Version::HTTP_2 => http2(),
        v => panic!("unsupported http version {:?}", v) // unreachable pattern lint triggered
    }
}
```

**Comment by seanmonstar** [comment]
Well shoot, that is unexpected. It was definitely hoped that by being an opaque struct, new versions could be added without making a breaking change. Has rustc always been able to do this? Or is it recent?

**Comment by carllerche** [comment]
Yeah, this sucks... but at least there won't be a new HTTP version added in the near term!

**Comment by kbacha** [comment]
https://github.com/hyperium/http/blob/master/src/version.rs#L43

Pretty sure it can tell based on the limited number of versions in the enum (thus there is a limited number of possible Version structs). I've seen crates that use an `__Reserved` enum before that is private:

```rust
/// Represents a version of the HTTP spec.
#[derive(PartialEq, PartialOrd, Copy, Clone, Eq, Ord, Hash)]
pub struct Version(Http);

impl Version {
    /// `HTTP/0.9`
    pub const HTTP_09: Version = Version(Http::Http09);

    /// `HTTP/1.0`
    pub const HTTP_10: Version = Version(Http::Http10);

    /// `HTTP/1.1`
    pub const HTTP_11: Version = Version(Http::Http11);

    /// `HTTP/2.0`
    pub const HTTP_2: Version = Version(Http::H2);
}

#[derive(PartialEq, PartialOrd, Copy, Clone, Eq, Ord, Hash)]
enum Http {
    Http09,
    Http10,
    Http11,
    H2,
    // Left unpublished to allow for future versions of http to occur
    __Reserved,
}
```

This should force you to implement a `_` or `v` since not all outputs are exposed. Of course, this still kinda requires that the versions exist in the crate and not the wild. Not sure which is better.

**Comment by seanmonstar** [comment]
Yes, the fix here is to add a variant. It would technically break anyone that has an exhaustive match on the version, so I've tossed it into the 0.2 milestone as a breaking change (shucks, the first one!).

**Comment by seanmonstar** [comment]
Thanks to #159, this is fixed in the unreleased 0.2.x branch!
